### PR TITLE
ARROW-12781: [R] Implement is.type() functions for dplyr

### PR DESCRIPTION
**Pull request by ianmcook**
- implements base R-flavored `is.*()` functions and rlang `is_*()` functions for checking column types in dplyr
- implements a general `is()` function like `methods::is()` for the same
- adds a `$type_id()` method to `Expression` and to some other R6 classes that were missing it
- changes how `Expression$type()`works, allowing the schema to be assigned to `Expression$schema` before
- adds support for the `where()` select helper in `relocate()` (but not in other verbs; that's ARROW-12778)

### CODE REVIEW

File: r/R/expression.R
Comment (ianmcook) [inline r/R/expression.R:86]: @nealrichardson I made this change because we need to call `Expression$type()` in contexts where we don't have access to `schema`.

Let me know if you'd prefer to retain `schema` as an _optional_ argument to `$type()` so that we can get the type either by specifying it in `schema` or passing it to `$bind()` before

Comment (ianmcook) [reply r/R/expression.R:86]: As you can see, calling the `$bind()` R6 method doesn't actually call the `Bind()` C++ function; I attempted that at first but ran into some problems so switched to doing it like this.

File: r/R/dplyr-functions.R
Comment (nealrichardson) [inline r/R/dplyr-functions.R:118]: I would spell these out too with the Type enum, just so they're obvious

File: r/R/dplyr-functions.R
Comment (nealrichardson) [inline r/R/dplyr-functions.R:130]: Technically `bit64::integer64` can't do uint64

```suggestion
  x$type_id() %in% Type["INT64"]
```

File: r/R/expression.R
Comment (nealrichardson) [inline r/R/expression.R:85]: How about this?

```suggestion
    type = function(schema = self$schema) {
      assert_that(!is.null(schema))
      compute___expr__type(self, schema)
```

I think it's misleading to have a `$bind()` method that doesn't actually Bind the Expression. I think you can achieve the same by doing `expr$schema <- schema` inside `relocate()` And also this way you retain the schema arg to `$type()` (so you don't need to modify the print method above)

File: r/R/dplyr-functions.R
Comment (nealrichardson) [inline r/R/dplyr-functions.R:144]: You may want to consider other method(s) to do the same logic but with Arrow-specific capabilities. Perhaps something as simple as this (following the base R signature)

```r
nse_funcs$is <- function(object, class2) {
  x$type$ToString() %in% class2
}
```

Also, are there hms/lubridate/etc. `is_*` functions, or tidyverse-spelling versions of these base R ones, to add?

**Review by nealrichardson — CHANGES_REQUESTED** [review]
Cool, thanks for picking this up

File: r/R/dplyr-functions.R
Comment (ianmcook) [reply r/R/dplyr-functions.R]: Done in 9f7fa6e67b12913876b166493eb7e9024e93b306

File: r/R/expression.R
Comment (ianmcook) [reply r/R/expression.R]: Done in 285f82606786d4ff850dbacc515c59466caada6d

I also did a little test and found out that it's 2x as fast to get the type ID by calling a C++ function dedicated for that purpose compared to calling the existing type function and getting the ID from it, so I implemented it that way.

Comment (ianmcook) [reply r/R/expression.R]: Just for posterity, here's the test code I ran
```r
library(dplyr)
library(arrow)

x <- as.data.frame(t(c(1:26)))
colnames(x) <- letters
x$z <- "z"
x <- Table$create(x)

start_time <- Sys.time()
for (i in 1:100) {
  q <- x %>% relocate(where(is.numeric), .after = where(is.character))
  w <- x %>% relocate(where(is.character), .before = where(is.numeric))
}
end_time <- Sys.time()
end_time - start_time

# ~8 seconds before change
# ~4 seconds after change
```

File: r/R/dplyr-functions.R
Comment (ianmcook) [reply r/R/dplyr-functions.R:144]: I implemented `is()` and the `rlang::is_*()` variants.

Implementing translations for the various date and time functions will require a lot more thought and planning. I think that's best done separately, in ARROW-12844. I noted that there that we should implement the date and time `is` and `as` functions.

File: r/R/dplyr-functions.R
Comment (ianmcook) [inline r/R/dplyr-functions.R:70]: @nealrichardson this canonicalizes type alias strings to match those returned by `DataType$ToString()`. It doesn't handle the ones that take parameters because that would get very complicated. Would this be useful for anywhere else? If so I can break it out into a separate function.

File: r/R/dplyr-functions.R
Comment (ianmcook) [inline r/R/dplyr-functions.R:166]: TODO: wait until #10375 is merged, rebase, simplify this to use `is.finite()`

File: r/R/dplyr-functions.R
Comment (ianmcook) [reply r/R/dplyr-functions.R]: This might be useful in the future so I broke it out into a separate function in 6f7c84b

File: r/R/type.R
Comment (nealrichardson) [inline r/R/type.R:427]: In some senses I think this is backwards: IIUC "float64" is canonical, but (for some reason) the C++ library prints it as "double".

File: r/R/type.R
Comment (nealrichardson) [inline r/R/type.R:426]: Should `as_type()` also wrap this and accept a string type name, returning a DataType? (I know I wrote the function basically to handle `double()`, but looking at the function name I'd kinda expect it to take the string name too.) If so, you could simplify `nse_funcs$is()`.

File: r/R/dplyr-functions.R
Comment (nealrichardson) [inline r/R/dplyr-functions.R:153]: This (and the others like it) won't work because `x` is an `Expression` and doesn't have a known `length`. I think we need to error if `!is.null(n)`

File: r/R/dplyr-functions.R
Comment (nealrichardson) [reply r/R/dplyr-functions.R:159]: Likewise we can't do that--`as.vector()` on an Expression in a dataset query isn't going to do what you hope

**Review by nealrichardson — CHANGES_REQUESTED** [review]

File: r/R/type.R
Comment (ianmcook) [reply r/R/type.R:427]: Per https://arrow.apache.org/docs/cpp/api/datatype.html it looks like "double" is canonical

File: r/R/type.R
Comment (ianmcook) [reply r/R/type.R:426]: That occurred to me too, and I like the idea if the TODOs below can be solved. I'll work on that.

File: r/R/dplyr-functions.R
Comment (ianmcook) [reply r/R/dplyr-functions.R:153]: D'oh, right, thanks.

File: r/R/type.R
Comment (nealrichardson) [reply r/R/type.R:427]: 🤷 Alrighty then, proceed

File: r/R/dplyr-functions.R
Comment (ianmcook) [reply r/R/dplyr-functions.R]: Fixed in 269bb24428c90971b0bd0de530d85c21229ec997

File: r/R/dplyr-functions.R
Comment (ianmcook) [reply r/R/dplyr-functions.R]: Fixed in 269bb24428c90971b0bd0de530d85c21229ec997

File: r/R/type.R
Comment (ianmcook) [reply r/R/type.R:431]: This turned out too be more complicated than anticipated, for two reasons:
- `canonical_type_str()` canonicalizes data type strings to match the output of `DataType$ToString()`, but if we wanted to accept data type strings in `as_type()`, we would instead need to convert them to the names of the R data type functions. This is basically like the opposite of what `canonical_type_str()` does so would require adding another function.
- For data types that take arguments, recognizing string representations of them and converting them to R data type function calls would be quite difficult.

For these reasons, I think we should save this `as_type()` improvement for another time.

Comment (ianmcook) [reply r/R/type.R:431]: ARROW-12908

**Pull request closed without merging**
